Feat new progressbar

lazy-val opened the pull request:
Progress bar consists now of 4 different coloured parts corresponding to the number of users that are in these four statuses: submitted, working, focused, idle.
The parts adjust their size when the numbers change.

Tests were added and improved.

What doesn't work: Receiving the events to get the initial users and the status updates.

Commit:
Changed the progress bar to let it display the amount of submissions, working users, focused users and idle users.
Added functions that will be called depending on the event.
TODO: Event handler, tests

Commit:
Demo progress bar

Commit:
Created the progress bar component.

Commit:
Fixed flex and positioning of the bars.

Commit:
Added the progress number also to the asa-progress component, some minor style changes, added a function to visualise a change given the duration.

Commit:
Added tests for asa-exercise-progress. TODO: Tests for the activity indication.

Commit:
Added tests to asa-exercise.

Commit:
Changed some indentation in the asq-exercise tests and fixed the multi-questions tests (rewriting as fixtures and fixing the error).

Commit:
Changed also the confidence tests to a fixture.

Commit:
Added event handler, commit to save changes.

Commit:
small corrections, (added forgotten parameter and missing comma)

Commit:
Changed the demo to also adjust the <asq-exercise-progress>.

Commit:
Added the functionality to visualise changes. Only receiving the events is missing.

triglian commented:
When the events work properly, don't forget to add docs for the functions.

Commit:
Progress bar changes now after receiving an event.
TODO: make effect, let the bar only change for the current exercise, write docs

Commit:
Variant 1: Progress bar only changes together with the effect.

Commit:
Variant 2: progress bar always changes directly, independently from the effect.

triglian commented on asq-confidence-style.html:
```diff
@@ -0,0 +1,103 @@
+      }
+
+      .rating > input[disabled] ~ label {
+          color: #666666;
```
use paper-colors?

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,310 @@
+    _userMadeTabHidden: function(user) {
+      this._makeSureUserExists(user);
+      this._users[user.id].idle = false;
+      this._notifyPolymer();
```
maybe use https://www.polymer-project.org/1.0/docs/devguide/model-data#notify-path ?

lazy-val replied:
I tried using the two suggested methods in the documentation (i.e using the polymer set function or the notify-path function), but it didn't seem to work, maybe because the parameters have to be a concatenated string, since I need to use variables in there.

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,310 @@
+    },
+
+    _subscribeToEvents: function(eventBus) {
+      let that = this;
```
instead of `that` you can bind the event hanlder function to `this` like so:

```js
eventBus.on('asq:sessionEventBeamer', function(evt) {

  const user = evt.data.sessionEvent.user;
  switch (evt.data.sessionEvent.type) {
    case 'folo-connected':
    case 'ghost-connected':
      return this._userConnected();
    //rest of your code
  }
}.bind(this))
```

Commit:
Added documentation to the progress bar and changed ‘that’ according to the pull request.
TODO: Look into the style of the asq-confidence

Commit:
Changed the colors in the asq-confidence to paper-colors.

Commit:
Some minor style changes to display the progress number in the same line as the progress bar.

Commit:
Changed some typos and corrected an issue that didn’t adjust the number of connected users when a user disconnected.

Commit:
Minor still fix to make the progress bar small again.

Commit:
Connected viewers shows no the correct value when the page is refreshed.

Commit:
Changed whiteListEntries back to user ids, to simplify debugging, because there were multiple whiteListEntries for the same user.
Added some logs for debugging.

Commit:
Removed the notifyPolymer function and used notifyPath instead. The number of submissions is now calculated by the asq-progressbar.

triglian commented on asq-exercise-progress-style.html:
```diff
@@ -0,0 +1,75 @@
+        --paper-progress-height: var(--asq-exercise-progress-height, 4px);
+      }
+
+      .background-progress paper-progress {
```
make it dynamic

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+<link rel="import" href="../asq-base/asq-base.html">
+<link rel="import" href="asq-exercise-progress-style.html">
+
+<dom-module id="asq-exercise-progress">
```
document style variables and mixing in this file.

triglian replied:
add a demo page just for <asq-exercise-progress.html> where you show code examples and results on how to change the style. Inspiration: https://www.webcomponents.org/element/PolymerElements/paper-button/demo/demo/index.html

triglian replied:
In JavaScript use only single quotes. In HTML use double quotes.

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+      }
+    </style>
+
+    <style include="asq-exercise-progress-style"></style>
```
You can merge this include to the previous <style> element

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+    </div>
+
+    <div id="progress-num">
+      <span>[[_submissionsNum]]</span>
```
Check if you can remove the unnecessary <span>s.

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+      * @type Number
+      * @default 0
+      */
+      connectedViewersNum: {
```
if it doesn't need to be public, add an underscore

triglian replied:
if `<asq-exercise>` doesn't need this, remove it from `<asq-exercise>`

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+      var viewers = this._viewers;
+      var that = this;
+      var num = Object.keys(viewers).reduce(function(acc, userId) {
+        if(viewers[userId].submitted == that.exUid) return ++acc;
```
refactor with `bind` so that you don't need the `var that = this;` assignment

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,509 @@
+
+    _subscribeToEvents: function(eventBus) {
+      eventBus.on('asq:connected-clients', this._onConnectedClients.bind(this));
+      eventBus.on('asq:sessionEventBeamer', function(evt) {
```
refactor the event handler to a named function like ` this._onConnectedClients.bind(this)`

triglian commented on asq-exercise.html:
```diff
@@ -6,14 +6,18 @@
 <!--
-`asq-exercise` groups ASQ `question types`. Every question should be be a child of an `asq-exercise` element. An `asq-exercise` element should conatin one ore more  `ASQ questions`. `asq-exercise` takes care of submitting and displaying progress for a group of questions.
+`asq-exercise` groups ASQ `question types`. Every question should be be a child
+of an `asq-exercise` element. An `asq-exercise` element should conatin one ore
```
"contain"

triglian commented on asq-exercise.html:
```diff
@@ -107,65 +129,44 @@
-      
-      /* no need to print the buttons 
+      /* no need to print the buttons
       fixme: when polymer support media queries for custom properties we can just apply a property to the footer and have the user provide the media query on their rules */
```
are media queries supported now?

triglian commented on asq-exercise.html:
```diff
@@ -6,14 +6,18 @@
 <link rel="import" href="../paper-progress/paper-progress.html">
```
check if there's functionality that is unnecessarily duplicated between `<asq-exercise>` and `asq-exercise-progress`

Commit:
- Changed “ in Javascript part to single quotes.
- asq-progressbar has now only one include for both stylesheets
- refactored with .bind(this) to not use this = that

Commit:
Refactored the session event handler to a named function.

Commit:
Added the documentation to asq-exercise-progress instead of asq-exercise.
Linked a demo page for the progress bar.

Commit:
Height of the background progress bar gets now adjusted dynamically.
Refactored some code.
Added the demo for the progress bar.

Commit:
Fixed a bug that prevented the activity animation from being displayed.

Commit:
Reverted a change that was done for testing purposes.

Commit:
Restores now the progress bar correctly when the presenter has to be restored.

Commit:
Presenter can now set a flag ‘directUpdate’ to decide if the progress bar updates together with the effect or directly.
Activity change effect works now together with setting the height of the background progress bar dynamically.

triglian commented on asq-exercise-progress.html:
```diff
@@ -0,0 +1,612 @@
+<link rel="import" href="../asq-base/asq-base.html">
+<link rel="import" href="asq-exercise-progress-style.html">
+
+<!--
```
Add some documentation on what this element does, what information it shows and maybe some design decisions. For example

> asq-exercise-progress is used to display the viewers' progress in answering an exercise. It consists of 4 partial progress bars: one that displays the total number of viewers that have submitted, one for ...

triglian committed:
add PolymerElements/iron-demo-helpers dep

triglian committed:
remove unused dep

triglian commented on demo/asq-exercise-progress-demo.html:
```diff
@@ -0,0 +1,110 @@
+
+  </head>
+
+  <body unresolved>
```
make the base font-size 14px;

Commit:
Corrected typo and added a description for the documentation of the asq exercise progress element.

triglian committed:
revised asq-exercise-progress docs

Commit:
Submissions are now displayed properly.

Commit:
Small bug fixes.

triglian committed:
fix typos

Commit:
Two fixes to make the tests (great) pass again.

Commit:
Merge branch 'feat-new-progressbar' of https://github.com/ASQ-USI-Elements/asq-exercise into feat-new-progressbar
